libraries/node-core-library/src/FileSystem.ts

```diff
@@ -0,0 +1,432 @@
+    }
+
+    contents = FileSystem._convertLineEndings(contents, options.convertLineEndings);
+
+    fsx.writeFileSync(filePath, contents, { encoding: options.encoding });
+  }
+
+  /**
+   * Reads the contents of a file into a string.
+   * Behind the scenes it uses `fs.readFileSync()`.
+   * @param filePath - The path of the file to be read.
+   * @param options - Optional settings that can change the behavior. Type: `IReadFileOptions`
+   */
+  public static readFile(filePath: string, options?: IReadFileOptions): string {
+    options = {
+      encoding: 'utf8',
+      convertLineEndings: NewlineConversion.None,
+      ...options
+    };
+
+    const contents: string = FileSystem.readFileToBuffer(filePath).toString(options.encoding);
+    return FileSystem._convertLineEndings(contents, options.convertLineEndings);
+  }
+
+  public static readFileToBuffer(filePath: string): Buffer {
+    return fsx.readFileSync(filePath);
+  }
+
+  /**
+   * Copies a file from one location to another.
+   * By default, destinationPath is overwritten if it already exists.
+   * Behind the scenes it uses `fs.copyFileSync()`.
+   * @param sourcePath - The absolute or relative path to the source file to be copied.
+   * @param destinationPath - The absolute or relative path to the new copy that will be created.
+   */
+  public static copyFile(sourcePath: string, destinationPath: string): void {
+    fsx.copySync(sourcePath, destinationPath);
+  }
+
+  /**
+   * Deletes a file. Can optionally throw if the file doesn't exist.
+   * Behind the scenes it uses `fs.unlinkSync()`.
+   * @param filePath - The absolute or relative path to the file that should be deleted.
+   * @param options - Optional settings that can change the behavior. Type: `IDeleteFileOptions`
+   */
+  public static deleteFile(filePath: string, options?: IDeleteFileOptions): void {
+    options = {
+      throwIfNotExists: false,
+      ...options
+    };
+
+    if (options.throwIfNotExists || FileSystem.exists(filePath)) {
+      fsx.unlinkSync(filePath);
+    }
+  }
+
+  // ===============
+  // LINK OPERATIONS
+  // ===============
+
+  /**
+   * Gets the statistics of a file, folder, or link. Does NOT follow the link to its target.
+   * Behind the scenes it uses `fs.lstatSync()`.
+   * @param path - The absolute or relative path to the file, folder, or link.
+   */
+  public static getLinkStatistics(path: string): fs.Stats {
+    return fsx.lstatSync(path);
+  }
+
+  /**
+   * Creates a symbolic link to a folder (on Windows this is a "directory junction").
+   * Behind the scenes it uses `fs.symlinkSync()`.
+   * @param linkSource - The absolute or relative path to the destination where the link should be created.
+   * @param linkTarget - The absolute or relative path to the target of the link.
+   */
+  public static createSymbolicLinkToFolder(linkSource: string, linkTarget: string): void {
+    // For directories, we use a Windows "junction".  On Unix, this produces a regular symlink.
```

Comment:
>"junction" [](start = 41, length = 10)

Windows has both directory junctions and symlinks:
[https://stackoverflow.com/questions/9042542/what-is-the-difference-between-ntfs-junction-points-and-symbolic-links](https://stackoverflow.com/questions/9042542/what-is-the-difference-between-ntfs-junction-points-and-symbolic-links)

Rush prefers junctions because they don't require admin access, however the person using the API should probably be asked to specify the type.

Because this is a cross-platform API, maybe it would be simplest to defer to NodeJS

Reply:
createSymlink --> file
createSymlinkDirectory --> directory
createSymlinkJunction --> junction

createHardLink


---
In reply to: [198671151](https://github.com/Microsoft/web-build-tools/pull/727#discussion_r198671151) [](ancestors = 198671151)